File: scripts/parse_chase_statements.py
#!/usr/bin/env python3
"""Parse Chase credit card statement PDFs and extract transactions.

Usage:
    python3 parse_chase_statements.py <pdf> [<pdf> ...]
    python3 parse_chase_statements.py --out transactions.xlsx ~/Downloads/Statements*.pdf

Output columns: file, date, merchant, amount, type, owner
    - date    : ISO date; year inferred from statement Opening/Closing Date
    - merchant: merchant name / transaction description
    - amount  : float; positive = purchase, negative = payment/credit
    - type    : PAYMENT / PURCHASE / CASH ADVANCE / FEE / INTEREST / BALANCE TRANSFER
    - owner   : "c" for Cavan-only purchases, "b" for shared purchases, blank for payments

Requires: pdfplumber, openpyxl (pip install pdfplumber openpyxl)
"""

from __future__ import annotations

import argparse
import csv
import re
import sys
from dataclasses import dataclass, asdict
from datetime import date, datetime
from pathlib import Path

import pdfplumber
from openpyxl import Workbook


PERIOD_RE = re.compile(
    r"Opening/Closing Date\s+"
    r"(\d{2}/\d{2}/\d{2})\s*-\s*(\d{2}/\d{2}/\d{2})"
)

SECTION_NAMES = [
    "PAYMENTS AND OTHER CREDITS",
    "PURCHASES",
    "PURCHASE",
    "CASH ADVANCES",
    "CASH ADVANCE",
    "FEES CHARGED",
    "INTEREST CHARGED",
    "BALANCE TRANSFERS",
]

# pdfplumber sometimes emits bold glyphs as doubled letters
# (e.g. "AACCCCOOUUNNTT AACCTTIIVVIITTYY"). Collapse any run of 2+ same chars
# to a single char for marker detection only.
_DUP_RE = re.compile(r"(.)\1+")


def _collapse_doubles(s: str) -> str:
    return _DUP_RE.sub(r"\1", s)


# Collapsed markers to test against; compare both raw and collapsed inputs.
_ACTIVITY_MARKERS = {"ACCOUNT ACTIVITY", _collapse_doubles("ACCOUNT ACTIVITY")}
_SECTION_MARKERS = {name: name for name in SECTION_NAMES}
_SECTION_MARKERS.update({_collapse_doubles(name): name for name in SECTION_NAMES})

# MM/DD <merchant...> <amount>
TX_RE = re.compile(
    r"^\s*(\d{2}/\d{2})\s+"
    r"(.+?)\s+"
    r"(-?\d[\d,]*\.\d{2})\s*$"
)

SECTION_TO_TYPE = {
    "PAYMENTS AND OTHER CREDITS": "PAYMENT",
    "PURCHASE": "PURCHASE",
    "PURCHASES": "PURCHASE",
    "CASH ADVANCE": "CASH ADVANCE",
    "CASH ADVANCES": "CASH ADVANCE",
    "FEES CHARGED": "FEE",
    "INTEREST CHARGED": "INTEREST",
    "BALANCE TRANSFERS": "BALANCE TRANSFER",
}

OWNER_C_PATTERNS = (
    re.compile(r"CHATGPT|OPENAI", re.IGNORECASE),
    re.compile(r"\bALO\b", re.IGNORECASE),
    re.compile(r"MAMMOTH", re.IGNORECASE),
    re.compile(r"SOLIDCORE", re.IGNORECASE),
    re.compile(r"CLASSPASS", re.IGNORECASE),
    re.compile(r"HIGHLINE", re.IGNORECASE),
)


@dataclass
class Transaction:
    file: str
    date: date
    merchant: str
    amount: float
    type: str
    owner: str


def _extract_text(pdf_path: Path) -> str:
    with pdfplumber.open(pdf_path) as pdf:
        return "\n".join((page.extract_text() or "") for page in pdf.pages)


def _statement_period(text: str) -> tuple[date, date]:
    m = PERIOD_RE.search(text)
    if not m:
        raise ValueError("Could not find Opening/Closing Date")
    open_d = datetime.strptime(m.group(1), "%m/%d/%y").date()
    close_d = datetime.strptime(m.group(2), "%m/%d/%y").date()
    return open_d, close_d


def _infer_year(mm: int, open_d: date, close_d: date) -> int:
    """Pick the correct calendar year for a MM/DD transaction.

    Statement periods straddle at most two calendar years (e.g. Dec->Jan).
    If the transaction month is >= opening month, use open year; else close year.
    """
    if open_d.year == close_d.year:
        return open_d.year
    return open_d.year if mm >= open_d.month else close_d.year


def parse_chase_statement(pdf_path: Path) -> list[Transaction]:
    text = _extract_text(pdf_path)
    open_d, close_d = _statement_period(text)

    records: list[Transaction] = []
    in_activity = False
    current_section: str | None = None

    for raw_line in text.splitlines():
        stripped = raw_line.strip()
        collapsed = _collapse_doubles(stripped)

        if stripped in _ACTIVITY_MARKERS or collapsed in _ACTIVITY_MARKERS:
            in_activity = True
            continue
        if not in_activity:
            continue

        # End of transaction area: YTD totals / interest tables.
        if (
            re.match(r"^20\d{2} Totals Year-to-Date", stripped)
            or stripped.startswith("INTEREST CHARGES")
            or collapsed.startswith("INTEREST CHARGES")
            or stripped.startswith("Totals Year-to-Date")
        ):
            in_activity = False
            continue

        if stripped in _SECTION_MARKERS:
            current_section = _SECTION_MARKERS[stripped]
            continue
        if collapsed in _SECTION_MARKERS:
            current_section = _SECTION_MARKERS[collapsed]
            continue

        tx_match = TX_RE.match(raw_line)
        if not tx_match:
            continue

        mmdd, merchant, amount_s = tx_match.groups()
        merchant = re.sub(r"\s+", " ", merchant).strip()
        # Some payment lines have a stray leading "& " artifact from the PDF.
        merchant = re.sub(r"^&\s+", "", merchant)

        mm, dd = int(mmdd[:2]), int(mmdd[3:5])
        year = _infer_year(mm, open_d, close_d)
        tx_date = date(year, mm, dd)
        amount = float(amount_s.replace(",", ""))

        tx_type = SECTION_TO_TYPE.get(current_section or "", "")
        owner = _classify_owner(merchant, amount, tx_type)

        records.append(
            Transaction(
                file=pdf_path.name,
                date=tx_date,
                merchant=merchant,
                amount=amount,
                type=tx_type,
                owner=owner,
            )
        )

    return records


def _classify_owner(merchant: str, amount: float, tx_type: str) -> str:
    if amount <= 0 or tx_type == "PAYMENT":
        return ""
    for pattern in OWNER_C_PATTERNS:
        if pattern.search(merchant):
            return "c"
    return "b"


def _summary_path(out_path: Path) -> Path:
    return out_path.with_name(f"{out_path.stem}_b_monthly_summary{out_path.suffix}")


def _monthly_b_summary(rows: list[Transaction]) -> list[dict[str, str | int | float]]:
    monthly: dict[str, dict[str, int | float | str]] = {}
    for row in rows:
        if row.owner != "b":
            continue
        month = row.date.strftime("%Y-%m")
        if month not in monthly:
            monthly[month] = {"month": month, "b_total": 0.0, "b_count": 0}
        monthly[month]["b_total"] += row.amount
        monthly[month]["b_count"] += 1

    return [
        {
            "month": month,
            "b_total": f"{values['b_total']:.2f}",
            "b_count": str(values["b_count"]),
            "c_share": f"{(values['b_total'] * 2 / 3):.2f}",
            "v_share": f"{(values['b_total'] / 3):.2f}",
        }
        for month, values in sorted(monthly.items())
    ]


def _metadata_rows(rows: list[Transaction], source_pdfs: list[Path]) -> list[dict[str, str]]:
    purchase_count = sum(1 for row in rows if row.amount > 0)
    payment_count = sum(1 for row in rows if row.amount <= 0)
    c_count = sum(1 for row in rows if row.owner == "c")
    b_count = sum(1 for row in rows if row.owner == "b")
    return [
        {"field": "generated_at", "value": datetime.now().isoformat(timespec="seconds")},
        {"field": "source_files", "value": ", ".join(pdf.name for pdf in source_pdfs)},
        {"field": "statement_file_count", "value": str(len(source_pdfs))},
        {"field": "transaction_row_count", "value": str(len(rows))},
        {"field": "purchase_row_count", "value": str(purchase_count)},
        {"field": "payment_row_count", "value": str(payment_count)},
        {"field": "owner_c_row_count", "value": str(c_count)},
        {"field": "owner_b_row_count", "value": str(b_count)},
        {
            "field": "owner_c_rules",
            "value": "CHATGPT/OPENAI, ALO, MAMMOTH, SOLIDCORE, CLASSPASS, HIGHLINE",
        },
        {"field": "shared_split", "value": "c=2/3, v=1/3"},
    ]


def _rows_to_dicts(rows: list[Transaction]) -> list[dict[str, str | float]]:
    return [
        {
            "file": row.file,
            "date": row.date.isoformat(),
            "merchant": row.merchant,
            "amount": row.amount,
            "type": row.type,
            "owner": row.owner,
        }
        for row in rows
    ]


def _write_csv(rows: list[Transaction], out_path: Path) -> None:
    with out_path.open("w", newline="") as fh:
        writer = csv.DictWriter(
            fh, fieldnames=["file", "date", "merchant", "amount", "type", "owner"]
        )
        writer.writeheader()
        writer.writerows(_rows_to_dicts(rows))


def _write_b_summary_csv(rows: list[Transaction], out_path: Path) -> Path:
    summary_path = _summary_path(out_path)
    summary_rows = _monthly_b_summary(rows)
    with summary_path.open("w", newline="") as fh:
        writer = csv.DictWriter(
            fh, fieldnames=["month", "b_total", "b_count", "c_share", "v_share"]
        )
        writer.writeheader()
        writer.writerows(summary_rows)
    return summary_path


def _autosize_worksheet(ws) -> None:
    for column_cells in ws.columns:
        values = [str(cell.value) if cell.value is not None else "" for cell in column_cells]
        max_len = max(len(value) for value in values) if values else 0
        ws.column_dimensions[column_cells[0].column_letter].width = min(max(max_len + 2, 10), 60)


def _write_xlsx(rows: list[Transaction], out_path: Path, source_pdfs: list[Path]) -> None:
    workbook = Workbook()

    tx_sheet = workbook.active
    tx_sheet.title = "Transactions"
    tx_fields = ["file", "date", "merchant", "amount", "type", "owner"]
    tx_sheet.append(tx_fields)
    for row in _rows_to_dicts(rows):
        tx_sheet.append([row[field] for field in tx_fields])

    summary_sheet = workbook.create_sheet("Monthly B Summary")
    summary_fields = ["month", "b_total", "b_count", "c_share", "v_share"]
    summary_sheet.append(summary_fields)
    for row in _monthly_b_summary(rows):
        summary_sheet.append([row[field] for field in summary_fields])

    metadata_sheet = workbook.create_sheet("Metadata")
    metadata_fields = ["field", "value"]
    metadata_sheet.append(metadata_fields)
    for row in _metadata_rows(rows, source_pdfs):
        metadata_sheet.append([row[field] for field in metadata_fields])

    _autosize_worksheet(tx_sheet)
    _autosize_worksheet(summary_sheet)
    _autosize_worksheet(metadata_sheet)
    workbook.save(out_path)


def _print_table(rows: list[Transaction]) -> None:
    if not rows:
        print("(no transactions)")
        return
    widths = {
        "file": max(len(r.file) for r in rows),
        "date": 10,
        "merchant": min(60, max(len(r.merchant) for r in rows)),
        "amount": 10,
        "type": max(len(r.type) for r in rows),
        "owner": 5,
    }
    header = (
        f"{'file':<{widths['file']}}  "
        f"{'date':<{widths['date']}}  "
        f"{'merchant':<{widths['merchant']}}  "
        f"{'amount':>{widths['amount']}}  "
        f"{'type':<{widths['type']}}  "
        f"{'owner':<{widths['owner']}}"
    )
    print(header)
    print("-" * len(header))
    for r in rows:
        merchant = r.merchant[: widths["merchant"]]
        print(
            f"{r.file:<{widths['file']}}  "
            f"{r.date.isoformat():<{widths['date']}}  "
            f"{merchant:<{widths['merchant']}}  "
            f"{r.amount:>{widths['amount']},.2f}  "
            f"{r.type:<{widths['type']}}  "
            f"{r.owner:<{widths['owner']}}"
        )


def main(argv: list[str]) -> int:
    p = argparse.ArgumentParser(
        description="Extract transactions from Chase credit card statement PDFs."
    )
    p.add_argument("pdfs", nargs="+", type=Path, help="Statement PDF files")
    p.add_argument(
        "--out",
        type=Path,
        help="Write results to .xlsx or .csv. If omitted, prints a table to stdout.",
    )
    args = p.parse_args(argv)

    all_tx: list[Transaction] = []
    for pdf in args.pdfs:
        if not pdf.exists():
            print(f"error: {pdf} does not exist", file=sys.stderr)
            return 1
        all_tx.extend(parse_chase_statement(pdf))

    all_tx.sort(key=lambda r: (r.date, r.file))

    if args.out:
        if args.out.suffix.lower() == ".xlsx":
            _write_xlsx(all_tx, args.out, args.pdfs)
            print(f"Wrote {len(all_tx)} transactions to workbook {args.out}")
            print("Workbook tabs: Transactions, Monthly B Summary, Metadata")
        else:
            _write_csv(all_tx, args.out)
            summary_path = _write_b_summary_csv(all_tx, args.out)
            print(f"Wrote {len(all_tx)} transactions to {args.out}")
            print(f"Wrote monthly shared summary to {summary_path}")
    else:
        _print_table(all_tx)

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: scripts/parse_chase_statements.R
#!/usr/bin/env Rscript

# Parse Chase credit card statement PDFs and extract transactions.
#
# Usage:
#   Rscript parse_chase_statements.R <pdf> [<pdf> ...]
#   Rscript parse_chase_statements.R --rules config/chase_owner_rules.csv --out transactions.xlsx ~/Downloads/Statements*.pdf
#
# Output columns: file, date, merchant, amount, type, owner
#   - date    : Date (inferred year from statement Opening/Closing Date)
#   - merchant: merchant name / transaction description
#   - amount  : numeric; positive = purchase, negative = payment/credit
#   - type    : "PAYMENT"/"PURCHASE"/"CASH ADVANCE"/"FEE"/"INTEREST" (section header)
#   - owner   : rule-matched owner; positive unmatched purchases default to "b"; payments blank

require_package <- function(package) {
  if (!requireNamespace(package, quietly = TRUE)) {
    stop("Install ", package, ": install.packages('", package, "')", call. = FALSE)
  }
}

empty_rules <- function() {
  data.frame(
    pattern = character(),
    owner = character(),
    label = character(),
    notes = character(),
    stringsAsFactors = FALSE
  )
}

read_owner_rules <- function(path = NULL) {
  require_package("readr")

  if (is.null(path) || !nzchar(path)) {
    return(empty_rules())
  }
  if (!file.exists(path)) {
    stop("Rules file does not exist: ", path, call. = FALSE)
  }

  rules <- readr::read_csv(path, col_types = readr::cols(.default = "c"), show_col_types = FALSE)
  required <- c("pattern", "owner", "label", "notes")
  missing <- setdiff(required, names(rules))
  if (length(missing) > 0) {
    stop("Rules file is missing columns: ", paste(missing, collapse = ", "), call. = FALSE)
  }

  rules <- as.data.frame(rules[required], stringsAsFactors = FALSE)
  rules$pattern <- trimws(rules$pattern)
  rules$owner <- trimws(rules$owner)
  rules <- rules[nzchar(rules$pattern), , drop = FALSE]
  invalid_owner <- !rules$owner %in% c("", "b", "c")
  if (any(invalid_owner)) {
    stop("Rules file owner values must be blank, 'b', or 'c'", call. = FALSE)
  }
  rules
}

classify_owner <- function(merchant, amount, type, owner_rules = empty_rules()) {
  if (is.na(amount) || amount <= 0 || identical(type, "PAYMENT")) {
    return("")
  }

  for (i in seq_len(nrow(owner_rules))) {
    if (grepl(owner_rules$pattern[[i]], merchant, ignore.case = TRUE)) {
      return(owner_rules$owner[[i]])
    }
  }

  "b"
}

summary_path <- function(out_file) {
  dir_name <- dirname(out_file)
  stem <- tools::file_path_sans_ext(basename(out_file))
  ext <- tools::file_ext(out_file)
  ext <- if (nzchar(ext)) paste0(".", ext) else ""
  file.path(dir_name, paste0(stem, "_b_monthly_summary", ext))
}

write_b_summary <- function(all_tx, out_file) {
  require_package("readr")

  summary_file <- summary_path(out_file)

  summary_table(all_tx) |>
    readr::write_csv(summary_file)

  summary_file
}

summary_table <- function(all_tx) {
  shared <- all_tx[!is.na(all_tx$owner) & all_tx$owner == "b", , drop = FALSE]
  if (nrow(shared) == 0) {
    return(data.frame(
      month = character(),
      b_total = numeric(),
      b_count = integer(),
      c_share = numeric(),
      v_share = numeric(),
      stringsAsFactors = FALSE
    ))
  }

  shared$month <- format(shared$date, "%Y-%m")
  totals <- aggregate(amount ~ month, shared, sum)
  counts <- aggregate(amount ~ month, shared, length)
  names(totals)[names(totals) == "amount"] <- "b_total"
  names(counts)[names(counts) == "amount"] <- "b_count"

  summary <- merge(totals, counts, by = "month", sort = TRUE)
  summary$b_total <- round(summary$b_total, 2)
  summary$c_share <- round(summary$b_total * 2 / 3, 2)
  summary$v_share <- round(summary$b_total / 3, 2)
  summary[, c("month", "b_total", "b_count", "c_share", "v_share")]
}

rule_summary <- function(owner_rules) {
  if (nrow(owner_rules) == 0) {
    return("none")
  }
  labels <- owner_rules$label[nzchar(owner_rules$label)]
  if (length(labels) == 0) {
    return(paste0(nrow(owner_rules), " rule(s)"))
  }
  paste(unique(labels), collapse = ", ")
}

metadata_table <- function(all_tx, source_files, owner_rules = empty_rules()) {
  data.frame(
    field = c(
      "generated_at",
      "source_files",
      "statement_file_count",
      "transaction_row_count",
      "purchase_row_count",
      "payment_row_count",
      "owner_c_row_count",
      "owner_b_row_count",
      "owner_rule_count",
      "owner_rule_labels",
      "shared_split"
    ),
    value = c(
      format(Sys.time(), "%Y-%m-%dT%H:%M:%S%z"),
      paste(basename(source_files), collapse = ", "),
      length(source_files),
      nrow(all_tx),
      sum(all_tx$amount > 0, na.rm = TRUE),
      sum(all_tx$amount <= 0, na.rm = TRUE),
      sum(all_tx$owner == "c", na.rm = TRUE),
      sum(all_tx$owner == "b", na.rm = TRUE),
      nrow(owner_rules),
      rule_summary(owner_rules),
      "c=2/3, v=1/3"
    ),
    stringsAsFactors = FALSE
  )
}

write_xlsx_workbook <- function(all_tx, out_file, source_files, owner_rules = empty_rules()) {
  require_package("openxlsx")

  wb <- openxlsx::createWorkbook()
  openxlsx::addWorksheet(wb, "Transactions")
  openxlsx::writeData(wb, "Transactions", all_tx)

  openxlsx::addWorksheet(wb, "Monthly B Summary")
  openxlsx::writeData(wb, "Monthly B Summary", summary_table(all_tx))

  openxlsx::addWorksheet(wb, "Metadata")
  openxlsx::writeData(wb, "Metadata", metadata_table(all_tx, source_files, owner_rules))

  openxlsx::saveWorkbook(wb, out_file, overwrite = TRUE)
}

parse_chase_statement_text <- function(text, source_file = "statement.txt", owner_rules = empty_rules()) {
  # Statement period: "Opening/Closing Date   MM/DD/YY - MM/DD/YY"
  period_match <- regmatches(
    text,
    regexpr(
      "Opening/Closing Date[[:space:]]+([0-9]{2}/[0-9]{2}/[0-9]{2})[[:space:]]*-[[:space:]]*([0-9]{2}/[0-9]{2}/[0-9]{2})",
      text
    )
  )
  if (length(period_match) == 0) {
    stop("Could not find Opening/Closing Date in ", source_file)
  }
  dates <- regmatches(
    period_match,
    gregexpr("[0-9]{2}/[0-9]{2}/[0-9]{2}", period_match)
  )[[1]]
  open_date <- as.Date(dates[1], format = "%m/%d/%y")
  close_date <- as.Date(dates[2], format = "%m/%d/%y")
  open_year <- as.integer(format(open_date, "%Y"))
  close_year <- as.integer(format(close_date, "%Y"))

  lines <- unlist(strsplit(text, "\n"))

  # Section headers that group transactions inside ACCOUNT ACTIVITY.
  section_regex <- paste0(
    "^[[:space:]]*(PAYMENTS AND OTHER CREDITS|PURCHASE[S]?|",
    "CASH ADVANCES?|FEES CHARGED|INTEREST CHARGED|",
    "BALANCE TRANSFERS)[[:space:]]*$"
  )

  # Transaction lines start with MM/DD and end with a $ amount.
  tx_regex <- paste0(
    "^[[:space:]]*([0-9]{2}/[0-9]{2})[[:space:]]+",
    "(.+?)[[:space:]]+",
    "(-?[0-9][0-9,]*\\.[0-9]{2})[[:space:]]*$"
  )

  in_activity <- FALSE
  current_section <- NA_character_
  records <- list()

  for (ln in lines) {
    stripped <- trimws(ln)

    if (grepl("^ACCOUNT ACTIVITY", stripped)) {
      in_activity <- TRUE
      next
    }
    if (!in_activity) next

    # End of transaction area: YTD totals, interest tables, etc.
    if (grepl("^20[0-9]{2} Totals Year-to-Date|^INTEREST CHARGES|^Totals Year-to-Date",
              stripped)) {
      in_activity <- FALSE
      next
    }

    if (grepl(section_regex, stripped)) {
      current_section <- sub("[[:space:]]+$", "", stripped)
      next
    }

    m <- regmatches(ln, regexec(tx_regex, ln))[[1]]
    if (length(m) == 4) {
      mmdd <- m[2]
      merchant <- trimws(gsub("[[:space:]]+", " ", m[3]))
      # Leading "& " appears on some payment lines; strip it.
      merchant <- sub("^& ", "", merchant)
      amount <- as.numeric(gsub(",", "", m[4]))

      # Year inference: if month >= opening month use open_year, else close_year.
      mm <- as.integer(substr(mmdd, 1, 2))
      yr <- if (mm >= as.integer(format(open_date, "%m"))) open_year else close_year
      # Handle calendar year rollover across Dec->Jan.
      if (open_year != close_year) {
        open_mm <- as.integer(format(open_date, "%m"))
        yr <- if (mm >= open_mm) open_year else close_year
      }
      tx_date <- as.Date(sprintf("%04d-%02d-%02d", yr, mm, as.integer(substr(mmdd, 4, 5))))

      type <- switch(
        toupper(current_section %||% ""),
        "PAYMENTS AND OTHER CREDITS" = "PAYMENT",
        "PURCHASE" = "PURCHASE",
        "PURCHASES" = "PURCHASE",
        "CASH ADVANCE" = "CASH ADVANCE",
        "CASH ADVANCES" = "CASH ADVANCE",
        "FEES CHARGED" = "FEE",
        "INTEREST CHARGED" = "INTEREST",
        "BALANCE TRANSFERS" = "BALANCE TRANSFER",
        NA_character_
      )
      owner <- classify_owner(merchant, amount, type, owner_rules)

      records[[length(records) + 1]] <- data.frame(
        file = basename(source_file),
        date = tx_date,
        merchant = merchant,
        amount = amount,
        type = type,
        owner = owner,
        stringsAsFactors = FALSE
      )
    }
  }

  if (length(records) == 0) {
    return(data.frame(
      file = character(), date = as.Date(character()),
      merchant = character(), amount = numeric(), type = character(), owner = character()
    ))
  }
  do.call(rbind, records)
}

parse_chase_statement <- function(pdf_path, owner_rules = empty_rules()) {
  require_package("pdftools")

  parse_chase_statement_text(
    paste(pdftools::pdf_text(pdf_path), collapse = "\n"),
    source_file = pdf_path,
    owner_rules = owner_rules
  )
}

`%||%` <- function(a, b) if (is.null(a) || is.na(a)) b else a

parse_cli_args <- function(args) {
  parsed <- list(out_file = NULL, rules_file = NULL, pdfs = character())
  i <- 1
  while (i <= length(args)) {
    arg <- args[[i]]
    if (arg == "--out") {
      if (i == length(args)) stop("--out requires a value", call. = FALSE)
      parsed$out_file <- args[[i + 1]]
      i <- i + 2
    } else if (arg == "--rules") {
      if (i == length(args)) stop("--rules requires a value", call. = FALSE)
      parsed$rules_file <- args[[i + 1]]
      i <- i + 2
    } else {
      parsed$pdfs <- c(parsed$pdfs, arg)
      i <- i + 1
    }
  }
  parsed
}

main <- function(args) {
  parsed <- parse_cli_args(args)
  if (length(parsed$pdfs) == 0) {
    cat("Usage: Rscript parse_chase_statements.R [--rules rules.csv] [--out out.xlsx] <pdf> [<pdf> ...]\n")
    quit(status = 1)
  }

  owner_rules <- read_owner_rules(parsed$rules_file)
  all_tx <- do.call(rbind, lapply(parsed$pdfs, parse_chase_statement, owner_rules = owner_rules))
  all_tx <- all_tx[order(all_tx$date, all_tx$file), ]

  if (!is.null(parsed$out_file)) {
    if (tolower(tools::file_ext(parsed$out_file)) == "xlsx") {
      write_xlsx_workbook(all_tx, parsed$out_file, parsed$pdfs, owner_rules)
      cat("Wrote", nrow(all_tx), "transactions to workbook", parsed$out_file, "\n")
      cat("Workbook tabs: Transactions, Monthly B Summary, Metadata\n")
    } else {
      utils::write.csv(all_tx, parsed$out_file, row.names = FALSE)
      summary_file <- write_b_summary(all_tx, parsed$out_file)
      cat("Wrote", nrow(all_tx), "transactions to", parsed$out_file, "\n")
      cat("Wrote monthly shared summary to", summary_file, "\n")
    }
  } else {
    # Print as a tidy table.
    old <- options(width = 200)
    on.exit(options(old), add = TRUE)
    print(all_tx, row.names = FALSE)
  }

  invisible(all_tx)
}

if (sys.nframe() == 0) {
  main(commandArgs(trailingOnly = TRUE))
}
